### PR TITLE
Provide more context when denying invalid type params 

### PR DESCRIPTION
**Pull request by estebank**

### CODE REVIEW

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 1]
```diff
@@ -1796,7 +1801,55 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                                    [.., segment, _] => (
+                                        segment.ident.span.shrink_to_hi().to(segment.args.map_or(
+                                            segment.ident.span.shrink_to_hi(),
+                                            |a| a.span_ext)),
+                                        false,
+                                    ),
+                                    [segment] => (
+                                        segment.ident.span.shrink_to_hi().to(segment.args.map_or(
+                                            segment.ident.span.shrink_to_hi(),
+                                            |a| a.span_ext)),
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs:1828]: Likewise.

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 2]
```diff
@@ -1796,7 +1801,55 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                                return;
+                            };
+                            let Some(type_name) = type_name else {
+                                err.note(&msg);
+                                return;
+                            };
+                            let suggestion = vec![
+                                if is_self {
+                                    // Account for people writing `Self::Variant::<Args>`, where
+                                    // `Self` is the enum.
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs:1844]: I don't understand the relation between this comment and the `match` above which defines `is_self`.

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 3]
```diff
@@ -2017,69 +2017,79 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         self.normalize_ty(span, tcx.mk_projection(item_def_id, item_substs))
     }
 
-    pub fn prohibit_generics<'a, T: IntoIterator<Item = &'a hir::PathSegment<'a>>>(
+    pub fn prohibit_generics<'a, T: Iterator<Item = &'a hir::PathSegment<'a>> + Clone>(
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs]: Isn't `T` always a slice `&[&hir::PathSegment<'_>]`? Can't we just take this?

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 4]
```diff
@@ -681,7 +681,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         trait_ref: &hir::TraitRef<'_>,
         self_ty: Ty<'tcx>,
     ) -> ty::TraitRef<'tcx> {
-        self.prohibit_generics(trait_ref.path.segments.split_last().unwrap().1);
+        self.prohibit_generics(trait_ref.path.segments.split_last().unwrap().1.iter());
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs]: Nit (pre-existing): an explicit slice may be easier to read.

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 5]
```diff
@@ -784,7 +784,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         let args = trait_segment.args();
         let infer_args = trait_segment.infer_args;
 
-        self.prohibit_generics(trait_ref.path.segments.split_last().unwrap().1);
+        self.prohibit_generics(trait_ref.path.segments.split_last().unwrap().1.iter());
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs]: Likewise.

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 6]
```diff
@@ -1796,7 +1801,55 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
                     if permit_variants {
                         tcx.check_stability(variant_def.def_id, Some(hir_ref_id), span, None);
-                        self.prohibit_generics(slice::from_ref(assoc_segment));
+                        self.prohibit_generics(slice::from_ref(assoc_segment).iter(), |err| {
+                            err.note("enum variants can't have type parameters");
+                            let type_name = tcx.opt_item_name(adt_def.did());
+                            let the_enum = type_name.map(|n| format!("enum `{n}`")).unwrap_or_else(|| "the enum".to_string());
+                            let msg = format!("you might have meant to specity type parameters on {the_enum}");
+                            let Some(args) = assoc_segment.args else { return; };
+                            let args_span = assoc_segment.ident.span.shrink_to_hi().to(args.span_ext);
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_typeck/src/astconv/mod.rs:1810]: I got it: you want the `::` too.  Could you add a comment?

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 7]
```diff
@@ -1796,7 +1801,55 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                            };
+                            let Some(type_name) = type_name else {
+                                err.note(&msg);
+                                return;
+                            };
+                            let suggestion = vec![
+                                if is_self {
+                                    // Account for people writing `Self::Variant::<Args>`, where
+                                    // `Self` is the enum.
+                                    (qself.span, format!("{type_name}{snippet}"))
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs:1845]: Should we have the same `TODO` here about checking that `type_name` accepts generics?

**Review by cjgillot — COMMENTED** [review]
The logic does not seem straightforward, so I have a lot of questions.

File: compiler/rustc_typeck/src/astconv/mod.rs (hunk 3 above)
Comment (estebank) [reply compiler/rustc_typeck/src/astconv/mod.rs]: I tried doing that. It isn't, there are a couple of callers that pass in an Iterator after skipping a few things. I think at least one uses `filter_map`, which isn't amenable to a direct subslicing approach.

File: compiler/rustc_typeck/src/astconv/mod.rs (hunk 2 above)
Comment (estebank) [reply compiler/rustc_typeck/src/astconv/mod.rs]: Expanded the comment

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 8]
```diff
@@ -2017,69 +2102,124 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                        let name = match res {
+                            Res::PrimTy(ty) => Some(ty.name()),
+                            Res::Def(_, def_id) => self.tcx().opt_item_name(def_id),
+                            _ => None,
+                        };
+                        Some((
+                            match name {
+                                Some(ty) => format!("{desc} `{ty}`"),
+                                None => desc.to_string(),
+                            },
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs:2132]: Should this be a single match on `res`?

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 9]
```diff
@@ -2017,69 +2102,124 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+            args.clone().fold((false, false, false, false), |(lt, ty, ct, inf), arg| match arg {
+                hir::GenericArg::Lifetime(_) => (true, ty, ct, inf),
+                hir::GenericArg::Type(_) => (lt, true, ct, inf),
+                hir::GenericArg::Const(_) => (lt, ty, true, inf),
+                hir::GenericArg::Infer(_) => (lt, ty, ct, true),
+            });
+        let mut emitted = false;
+        if lt || ty || ct || inf {
+            let arg_spans: Vec<Span> = args
+                .map(|arg| match arg {
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs:2162]: This is `arg.span()`.

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 10]
```diff
@@ -2017,69 +2102,124 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                    hir::GenericArg::Lifetime(lt) => lt.span,
+                    hir::GenericArg::Type(ty) => ty.span,
+                    hir::GenericArg::Const(ct) => ct.span,
+                    hir::GenericArg::Infer(inf) => inf.span,
+                })
+                .collect();
+
+            let mut types = Vec::with_capacity(4);
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs:2170]:
```suggestion
            let mut kinds = Vec::with_capacity(4);
```

File: compiler/rustc_typeck/src/astconv/mod.rs (hunk 8 above)
Comment (estebank) [reply compiler/rustc_typeck/src/astconv/mod.rs:2132]: It would further complicate the match because the `Def` might not have an item name, right?

Comment (estebank) [reply compiler/rustc_typeck/src/astconv/mod.rs:2127]: changed

**Pull request merged**